### PR TITLE
Refactor sphinx.theming (T/O)

### PR DESCRIPTION
**Pull request by tk0miya**

### CODE REVIEW

File: sphinx/theming.py [hunk 1]
```diff
@@ -11,212 +11,143 @@
...
-            self.base = Theme(inherit)
+            return [self.themedir] + self.base.get_dirchain()
 
-    def get_confstr(self, section, name, default=NODEFAULT):
+    def get_dirchain(self):
+        # type: () -> List[unicode]
+        """Return a list of theme directories, beginning with this theme's,
+        then the base theme's, then that one's base theme's, etc.
+        """
+        return self.dirs
```
Comment (shimizukawa) [inline sphinx/theming.py]: I think this is a backward-compatibility method. However, this Theme class is used only from Sphinx inside. IMO, this compat method is not needed.

File: sphinx/theming.py [hunk 2]
```diff
@@ -11,212 +11,143 @@
...
-                raise ThemeError('setting %s.%s occurs in none of the '
-                                 'searched theme configs' % (section, name))
+                raise ThemeError(_('setting %s.%s occurs in none of the '
+                                   'searched theme configs') % (section, name))
             else:
                 return default
 
-    def get_options(self, overrides):
-        # type: (Dict) -> Any
+    def get_confstr(self, section, name, default=NODEFAULT):
```
Comment (shimizukawa) [inline sphinx/theming.py]: This compat method should be removed too.

File: sphinx/theming.py [hunk 3]
```diff
@@ -11,212 +11,143 @@
...
+
+        self.config = configparser.RawConfigParser()
+        self.config.read(path.join(self.themedir, THEMECONF))  # type: ignore
 
         try:
-            inherit = self.themeconf.get('theme', 'inherit')
+            inherit = self.config.get('theme', 'inherit')
         except configparser.NoOptionError:
-            raise ThemeError('theme %r doesn\'t have "inherit" setting' % name)
+            raise ThemeError(_('theme %r doesn\'t have "inherit" setting') % name)
```
Comment (shimizukawa) [inline sphinx/theming.py:76]: I think that ERROR messages should be output in the console's language setting.

Sphinx's `_(...)` determines the language according to the conf.py language setting. Therefore, when Klingon language document is being built, error messages will be displayed in Klingon language whatever the console language is. (I can't read it)

File: tests/util.py [hunk 4]
```diff
@@ -168,7 +167,6 @@ def __init__(self, buildername='html', testroot=None, srcdir=None,
             raise
 
     def cleanup(self, doctrees=False):
-        Theme.themes.clear()
```
Comment (shimizukawa) [inline tests/util.py:171]: Woooooo!!!!! Very nice!!

**Review by shimizukawa — APPROVED** [review]
LGTM with nits ;)
👍 👍 👍

File: sphinx/theming.py (hunk 3 above)
Comment (tk0miya) [reply sphinx/theming.py:76]: Either is fine, but sphinx already uses i18n texts for error and warning messages.
So we have to do that on another issue.
What do you think? Should I revert this ASAP?

Comment (shimizukawa) [reply sphinx/theming.py:76]: For now, ether OK to me too.
In the future, let's change to output in console language.

File: sphinx/theming.py (hunk 1 above)
Comment (tk0miya) [reply sphinx/theming.py]: Removed

**Comment by tk0miya** [comment]
Merged.
@shimizukawa Thank you for reviewing!

**Pull request merged**